File: go/internal/decon/patterns/patterns.go
// Package patterns ports evm_decon/patterns.py: proxy detect, security flags,
// embedded string scan, opcode-frequency stats.
package patterns

import (
	"encoding/hex"
	"sort"
	"strings"

	"github.com/evm-auditor/evm-auditor/internal/decon/disasm"
)

// EIP-1167 minimal-proxy template constants.
const (
	eip1167Prefix      = "363d3d373d3d3d363d73"
	eip1167Suffix      = "5af43d82803e903d91602b57fd5bf3"
	eip1167TotalBytes  = 45
	eip1967ImplSlot    = "360894a13ba1a3210667c828492db98dca3e2076cc3735a920a3ca505d382bbc"
	eip1967AdminSlot   = "b53127684a568b3173ae13b9f8a6016e243e63b6e8ee1178d6a717850b5d6103"
	eip1967BeaconSlot  = "a3f0ad74e5423aebfd80d3ef4346578335a9a72aeaee59ff6cb3582b35133d50"
)

// ProxyInfo mirrors evm_decon.patterns.ProxyInfo.
type ProxyInfo struct {
	IsProxy               bool
	ProxyType             string
	ImplementationAddress string
	Details               map[string]any
}

// SecurityFlags mirrors evm_decon.patterns.SecurityFlags.
type SecurityFlags struct {
	HasSelfdestruct      bool
	HasDelegatecall      bool
	HasCallcode          bool
	HasCreate            bool
	HasCreate2           bool
	HasOrigin            bool
	HasStaticcall        bool
	HasSendOrTransfer    bool
	SelfdestructOffsets  []int
	DelegatecallOffsets  []int
}

// StringLiteral mirrors evm_decon.patterns.StringLiteral.
type StringLiteral struct {
	Offset   int
	Value    string
	Encoding string
	Length   int
}

// Result mirrors evm_decon.patterns.PatternResult. Standards detection is left
// out (port pending — needs known_signatures table).
type Result struct {
	Proxy             ProxyInfo
	Security          SecurityFlags
	Standards         []map[string]any
	Strings           []StringLiteral
	OpcodeFrequency   []KV
	OpcodeCategories  []KV
}

// KV is a single (opcode, count) pair preserved in descending-count order to
// mirror Python's `dict(sorted(... key=-count))`.
type KV struct {
	Key   string
	Value int
}

// Detect runs the full pattern pipeline.
func Detect(d disasm.Result, bytecodeHex string, _ map[string]struct{}) Result {
	return Result{
		Proxy:            detectProxy(d, bytecodeHex),
		Security:         detectSecurity(d),
		Standards:        []map[string]any{}, // TODO: port known_signatures.detect_standards
		Strings:          extractStrings(bytecodeHex),
		OpcodeFrequency:  opcodeFrequency(d),
		OpcodeCategories: categoryFrequency(d),
	}
}

func detectProxy(d disasm.Result, bytecodeHex string) ProxyInfo {
	hexStr := strings.TrimPrefix(strings.ToLower(strings.TrimSpace(bytecodeHex)), "0x")
	bc, err := hex.DecodeString(hexStr)
	if err != nil {
		return ProxyInfo{}
	}

	// 1. EIP-1167 exact match
	if len(bc) == eip1167TotalBytes {
		prefix, _ := hex.DecodeString(eip1167Prefix)
		suffix, _ := hex.DecodeString(eip1167Suffix)
		if bytesEqual(bc[:10], prefix) && bytesEqual(bc[30:], suffix) {
			return ProxyInfo{
				IsProxy:               true,
				ProxyType:             "EIP-1167",
				ImplementationAddress: "0x" + hex.EncodeToString(bc[10:30]),
				Details:               map[string]any{"pattern": "minimal_proxy", "exact_match": true},
			}
		}
	}

	// 1b. EIP-1167 embedded
	if idx := strings.Index(hexStr, eip1167Prefix); idx >= 0 {
		addrStart := idx + 20
		suffixStart := addrStart + 40
		if suffixStart+len(eip1167Suffix) <= len(hexStr) &&
			hexStr[suffixStart:suffixStart+len(eip1167Suffix)] == eip1167Suffix {
			return ProxyInfo{
				IsProxy:               true,
				ProxyType:             "EIP-1167",
				ImplementationAddress: "0x" + hexStr[addrStart:addrStart+40],
				Details:               map[string]any{"pattern": "minimal_proxy", "embedded": true},
			}
		}
	}

	// 2. EIP-1967
	hasEIP1967 := false
	for _, ins := range d.Instructions {
		if ins.Mnemonic != "PUSH32" || ins.Operand == nil {
			continue
		}
		operand := ins.Operand.Text(16)
		// Pad to 64 chars for comparison.
		for len(operand) < 64 {
			operand = "0" + operand
		}
		if operand == eip1967ImplSlot || operand == eip1967AdminSlot || operand == eip1967BeaconSlot {
			hasEIP1967 = true
			break
		}
	}
	if hasEIP1967 && hasOpcode(d, "DELEGATECALL") {
		return ProxyInfo{
			IsProxy:   true,
			ProxyType: "EIP-1967",
			Details:   map[string]any{"pattern": "upgradeable_proxy", "slot_detected": true},
		}
	}

	// 3. Generic small-delegatecall heuristic
	if hasOpcode(d, "DELEGATECALL") && d.BytecodeSize < 500 {
		return ProxyInfo{
			IsProxy:   true,
			ProxyType: "DELEGATECALL-based",
			Details:   map[string]any{"pattern": "generic_proxy", "heuristic": true},
		}
	}
	return ProxyInfo{}
}

func detectSecurity(d disasm.Result) SecurityFlags {
	flags := SecurityFlags{}
	for i, ins := range d.Instructions {
		switch ins.Mnemonic {
		case "SELFDESTRUCT":
			flags.HasSelfdestruct = true
			flags.SelfdestructOffsets = append(flags.SelfdestructOffsets, ins.Offset)
		case "DELEGATECALL":
			flags.HasDelegatecall = true
			flags.DelegatecallOffsets = append(flags.DelegatecallOffsets, ins.Offset)
		case "CALLCODE":
			flags.HasCallcode = true
		case "CREATE":
			flags.HasCreate = true
		case "CREATE2":
			flags.HasCreate2 = true
		case "ORIGIN":
			flags.HasOrigin = true
		case "STATICCALL":
			flags.HasStaticcall = true
		case "CALL":
			if i >= 1 {
				prev := d.Instructions[i-1]
				if (prev.Mnemonic == "PUSH2" || prev.Mnemonic == "PUSH3") && prev.Operand != nil &&
					prev.Operand.Int64() == 2300 {
					flags.HasSendOrTransfer = true
				}
			}
		}
	}
	return flags
}

func extractStrings(bytecodeHex string) []StringLiteral {
	hexStr := strings.TrimPrefix(strings.ToLower(strings.TrimSpace(bytecodeHex)), "0x")
	bc, err := hex.DecodeString(hexStr)
	if err != nil {
		return nil
	}
	out := []StringLiteral{}
	current := []byte{}
	start := 0
	flush := func() {
		if len(current) >= 6 {
			text := string(current)
			if !allHexChars(strings.ToLower(text)) {
				out = append(out, StringLiteral{
					Offset: start, Value: text, Encoding: "ascii", Length: len(text),
				})
			}
		}
		current = current[:0]
	}
	for i, b := range bc {
		if b >= 0x20 && b <= 0x7e {
			if len(current) == 0 {
				start = i
			}
			current = append(current, b)
		} else {
			flush()
		}
	}
	flush()
	return out
}

func opcodeFrequency(d disasm.Result) []KV {
	counts := map[string]int{}
	for _, ins := range d.Instructions {
		counts[ins.Mnemonic]++
	}
	return sortedDesc(counts)
}

func categoryFrequency(d disasm.Result) []KV {
	counts := map[string]int{}
	for _, ins := range d.Instructions {
		counts[disasm.Lookup(ins.Opcode).Category]++
	}
	return sortedDesc(counts)
}

func sortedDesc(m map[string]int) []KV {
	out := make([]KV, 0, len(m))
	for k, v := range m {
		out = append(out, KV{Key: k, Value: v})
	}
	sort.SliceStable(out, func(i, j int) bool {
		if out[i].Value != out[j].Value {
			return out[i].Value > out[j].Value
		}
		return out[i].Key < out[j].Key
	})
	return out
}

func hasOpcode(d disasm.Result, op string) bool {
	for _, ins := range d.Instructions {
		if ins.Mnemonic == op {
			return true
		}
	}
	return false
}

func bytesEqual(a, b []byte) bool {
	if len(a) != len(b) {
		return false
	}
	for i := range a {
		if a[i] != b[i] {
			return false
		}
	}
	return true
}

func allHexChars(s string) bool {
	for _, c := range s {
		if !((c >= '0' && c <= '9') || (c >= 'a' && c <= 'f')) {
			return false
		}
	}
	return true
}

File: evm_decon/patterns.py
"""
Pattern detector for EVM bytecode.

Detects proxy patterns, security-relevant opcodes, ERC standard compliance,
and other structural patterns from disassembled bytecode.
"""

from __future__ import annotations

from dataclasses import dataclass, field
from typing import Optional
from .disassembler import Instruction, DisassemblyResult
from .known_signatures import detect_standards


# EIP-1167 Minimal Proxy template (without the 20-byte address)
_EIP1167_PREFIX = bytes.fromhex("363d3d373d3d3d363d73")
_EIP1167_SUFFIX = bytes.fromhex("5af43d82803e903d91602b57fd5bf3")
_EIP1167_TOTAL_LENGTH = 45  # 10 + 20 + 15

# EIP-1967 implementation storage slot
_EIP1967_IMPL_SLOT = "360894a13ba1a3210667c828492db98dca3e2076cc3735a920a3ca505d382bbc"
_EIP1967_ADMIN_SLOT = "b53127684a568b3173ae13b9f8a6016e243e63b6e8ee1178d6a717850b5d6103"
_EIP1967_BEACON_SLOT = "a3f0ad74e5423aebfd80d3ef4346578335a9a72aeaee59ff6cb3582b35133d50"


@dataclass
class ProxyInfo:
    is_proxy: bool
    proxy_type: Optional[str]  # "EIP-1167", "EIP-1967", "DELEGATECALL-based", None
    implementation_address: Optional[str]  # for EIP-1167
    details: dict = field(default_factory=dict)


@dataclass
class SecurityFlags:
    has_selfdestruct: bool
    has_delegatecall: bool
    has_callcode: bool
    has_create: bool
    has_create2: bool
    has_origin: bool       # tx.origin usage
    has_staticcall: bool
    has_send_or_transfer: bool  # heuristic
    selfdestruct_offsets: list[int] = field(default_factory=list)
    delegatecall_offsets: list[int] = field(default_factory=list)


@dataclass
class StringLiteral:
    offset: int
    value: str
    encoding: str  # "utf8", "ascii"
    length: int


@dataclass
class PatternResult:
    proxy: ProxyInfo
    security: SecurityFlags
    standards: list[dict]
    strings: list[StringLiteral]
    opcode_frequency: dict[str, int]
    opcode_categories: dict[str, int]  # category → count


def detect_patterns(
    disasm: DisassemblyResult,
    bytecode_hex: str,
    selector_set: set[str],
) -> PatternResult:
    """
    Run all pattern detectors on disassembled bytecode.

    Args:
        disasm: Disassembly result
        bytecode_hex: Original hex bytecode (for raw pattern matching)
        selector_set: Set of extracted function selectors (hex, no 0x prefix)
    """
    proxy = _detect_proxy(disasm, bytecode_hex)
    security = _detect_security_flags(disasm)
    standards = detect_standards(selector_set)
    strings = _extract_strings(bytecode_hex)
    opcode_freq = _compute_opcode_frequency(disasm)
    opcode_cats = _compute_category_frequency(disasm)

    return PatternResult(
        proxy=proxy,
        security=security,
        standards=standards,
        strings=strings,
        opcode_frequency=opcode_freq,
        opcode_categories=opcode_cats,
    )


def _detect_proxy(disasm: DisassemblyResult, bytecode_hex: str) -> ProxyInfo:
    """Detect proxy contract patterns."""

    hex_str = bytecode_hex.strip().lower()
    if hex_str.startswith("0x"):
        hex_str = hex_str[2:]

    try:
        bytecode = bytes.fromhex(hex_str)
    except ValueError:
        return ProxyInfo(is_proxy=False, proxy_type=None, implementation_address=None)

    # 1. EIP-1167 Minimal Proxy
    if len(bytecode) == _EIP1167_TOTAL_LENGTH:
        if (
            bytecode[:10] == _EIP1167_PREFIX
            and bytecode[30:] == _EIP1167_SUFFIX
        ):
            impl_addr = "0x" + bytecode[10:30].hex()
            return ProxyInfo(
                is_proxy=True,
                proxy_type="EIP-1167",
                implementation_address=impl_addr,
                details={"pattern": "minimal_proxy", "exact_match": True},
            )

    # Also check for EIP-1167 embedded in larger bytecode
    eip1167_idx = hex_str.find("363d3d373d3d3d363d73")
    if eip1167_idx >= 0:
        # Check if the suffix follows 20 bytes later
        addr_start = eip1167_idx + 20  # 10 bytes * 2 hex chars
        suffix_start = addr_start + 40  # 20 bytes * 2 hex chars
        expected_suffix = "5af43d82803e903d91602b57fd5bf3"
        if hex_str[suffix_start:suffix_start + len(expected_suffix)] == expected_suffix:
            impl_addr = "0x" + hex_str[addr_start:addr_start + 40]
            return ProxyInfo(
                is_proxy=True,
                proxy_type="EIP-1167",
                implementation_address=impl_addr,
                details={"pattern": "minimal_proxy", "embedded": True},
            )

    # 2. EIP-1967 storage slot pattern
    has_eip1967 = False
    for inst in disasm.instructions:
        if inst.opcode == "PUSH32" and inst.operand:
            operand = inst.operand.lower().replace("0x", "")
            if operand == _EIP1967_IMPL_SLOT:
                has_eip1967 = True
                break
            if operand == _EIP1967_ADMIN_SLOT:
                has_eip1967 = True
                break
            if operand == _EIP1967_BEACON_SLOT:
                has_eip1967 = True
                break

    if has_eip1967:
        # Check for DELEGATECALL
        has_delegatecall = any(i.opcode == "DELEGATECALL" for i in disasm.instructions)
        if has_delegatecall:
            return ProxyInfo(
                is_proxy=True,
                proxy_type="EIP-1967",
                implementation_address=None,  # stored in slot, can't read from bytecode alone
                details={"pattern": "upgradeable_proxy", "slot_detected": True},
            )

    # 3. Generic DELEGATECALL-based proxy heuristic
    # If contract is small and has DELEGATECALL, it's likely a proxy
    has_delegatecall = any(i.opcode == "DELEGATECALL" for i in disasm.instructions)
    if has_delegatecall and disasm.bytecode_size < 500:
        return ProxyInfo(
            is_proxy=True,
            proxy_type="DELEGATECALL-based",
            implementation_address=None,
            details={"pattern": "generic_proxy", "heuristic": True},
        )

    return ProxyInfo(is_proxy=False, proxy_type=None, implementation_address=None)


def _detect_security_flags(disasm: DisassemblyResult) -> SecurityFlags:
    """Detect security-relevant opcodes."""

    selfdestruct_offsets = []
    delegatecall_offsets = []
    flags = {
        "selfdestruct": False,
        "delegatecall": False,
        "callcode": False,
        "create": False,
        "create2": False,
        "origin": False,
        "staticcall": False,
        "send_transfer": False,
    }

    for inst in disasm.instructions:
        if inst.opcode == "SELFDESTRUCT":
            flags["selfdestruct"] = True
            selfdestruct_offsets.append(inst.offset)
        elif inst.opcode == "DELEGATECALL":
            flags["delegatecall"] = True
            delegatecall_offsets.append(inst.offset)
        elif inst.opcode == "CALLCODE":
            flags["callcode"] = True
        elif inst.opcode == "CREATE":
            flags["create"] = True
        elif inst.opcode == "CREATE2":
            flags["create2"] = True
        elif inst.opcode == "ORIGIN":
            flags["origin"] = True
        elif inst.opcode == "STATICCALL":
            flags["staticcall"] = True
        elif inst.opcode == "CALL":
            # Heuristic: if CALL is preceded by PUSH with gas=2300, it's send/transfer
            idx = disasm.instructions.index(inst)
            if idx >= 1:
                prev = disasm.instructions[idx - 1]
                if prev.opcode in ("PUSH2", "PUSH3") and prev.operand_value == 2300:
                    flags["send_transfer"] = True

    return SecurityFlags(
        has_selfdestruct=flags["selfdestruct"],
        has_delegatecall=flags["delegatecall"],
        has_callcode=flags["callcode"],
        has_create=flags["create"],
        has_create2=flags["create2"],
        has_origin=flags["origin"],
        has_staticcall=flags["staticcall"],
        has_send_or_transfer=flags["send_transfer"],
        selfdestruct_offsets=selfdestruct_offsets,
        delegatecall_offsets=delegatecall_offsets,
    )


def _extract_strings(bytecode_hex: str) -> list[StringLiteral]:
    """
    Extract embedded string literals from bytecode.

    Solidity stores string literals as PUSH32 sequences or in memory.
    We use a heuristic: look for sequences of printable ASCII bytes >= 4 chars.
    """
    hex_str = bytecode_hex.strip().lower()
    if hex_str.startswith("0x"):
        hex_str = hex_str[2:]

    try:
        raw = bytes.fromhex(hex_str)
    except ValueError:
        return []

    strings = []
    current = bytearray()
    start_offset = 0

    for i, b in enumerate(raw):
        if 0x20 <= b <= 0x7e:  # printable ASCII
            if not current:
                start_offset = i
            current.append(b)
        else:
            if len(current) >= 6:  # minimum 6 chars to avoid noise
                text = current.decode("ascii")
                # Filter out hex-looking strings and obvious noise
                if not all(c in "0123456789abcdef" for c in text.lower()):
                    strings.append(StringLiteral(
                        offset=start_offset,
                        value=text,
                        encoding="ascii",
                        length=len(text),
                    ))
            current = bytearray()

    # Don't forget trailing
    if len(current) >= 6:
        text = current.decode("ascii")
        if not all(c in "0123456789abcdef" for c in text.lower()):
            strings.append(StringLiteral(
                offset=start_offset,
                value=text,
                encoding="ascii",
                length=len(text),
            ))

    return strings


def _compute_opcode_frequency(disasm: DisassemblyResult) -> dict[str, int]:
    """Compute frequency of each opcode."""
    freq: dict[str, int] = {}
    for inst in disasm.instructions:
        freq[inst.opcode] = freq.get(inst.opcode, 0) + 1
    return dict(sorted(freq.items(), key=lambda x: -x[1]))


def _compute_category_frequency(disasm: DisassemblyResult) -> dict[str, int]:
    """Compute frequency by opcode category."""
    freq: dict[str, int] = {}
    for inst in disasm.instructions:
        cat = inst.info.category
        freq[cat] = freq.get(cat, 0) + 1
    return dict(sorted(freq.items(), key=lambda x: -x[1]))
